evmcore/state_processor.go

```diff
@@ -101,17 +104,48 @@ func (p *StateProcessor) Process(
 		if err != nil {
 			return nil, nil, nil, fmt.Errorf("could not apply tx %d [%v]: %w", i, tx.Hash().Hex(), err)
 		}
-		if substate.RecordReplay {
+		if innerSubstate.RecordReplay {
 			// save tx substate into DBs, merge block hashes to env
 			etherBlock := block.RecordingEthBlock()
+			to := stypes.Address(msg.To().Bytes())
+			dataHash := hash.Keccak256Hash(msg.Data())
 			recording := substate.NewSubstate(
 				statedb.SubstatePreAlloc,
 				statedb.SubstatePostAlloc,
-				substate.NewSubstateEnv(etherBlock, statedb.SubstateBlockHashes),
-				substate.NewSubstateMessage(&msg),
-				substate.NewSubstateResult(receipt),
+				substate.NewEnv(
+					stypes.Address(etherBlock.Coinbase()),
+					etherBlock.Difficulty(),
+					etherBlock.GasLimit(),
+					etherBlock.NumberU64(),
+					etherBlock.Time(),
+					etherBlock.BaseFee(),
+					innerSubstate.HashGethToSubstate(statedb.SubstateBlockHashes)),
+				substate.NewMessage(
+					msg.Nonce(),
+					msg.IsFake(),
+					msg.GasPrice(),
+					msg.Gas(),
+					stypes.Address(msg.From()),
+					&to,
+					msg.Value(),
+					msg.Data(),
+					&dataHash,
+					innerSubstate.AccessListGethToSubstate(msg.AccessList()),
+					msg.GasFeeCap(),
+					msg.GasTipCap()),
+				substate.NewResult(
+					receipt.Status,
+					receipt.Bloom.Bytes(),
+					innerSubstate.LogsGethToSubstate(receipt.Logs),
+					stypes.Address(receipt.ContractAddress),
+					receipt.GasUsed),
```

Comment:
Each New function takes long arguments. I want to suggest moving them to the `substate` package.